### PR TITLE
Closes 5401: Prevent useless requests when RUCSS and Preload activated

### LINKED ISSUE #5401: Guard against unnecessary request when RUCSS and Preload are used together
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
When RUCSS and Preload are used together it's possible that we'll do one additional, unnecessary request to the URL.

**To Reproduce**
1. Enable Preload and RUCSS, make sure that the Cache and Used CSS are not present yet
2. Preload will start triggering Used CSS requests to SaaS
3. Click on the `Clear and Preload` the Cache
4. Pages that are waiting for the Used CSS will be preloaded
5. When we receive the Used CSS, they'll be cleared and preloaded once again to inject the Used CSS

**Expected behavior**
Pages that are waiting for the Used CSS should not be preloaded.

**Additional context**
That's not very likely scenario to happen but it's good to save as many resources as possible.

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

### PR DESCRIPTION
**Pull request by CrochetFeve0251**
## Description
Fix a problem where WP Rocket was preloading pages multiple times when RUCSS was activated.

For that we blocked preload from pages that are loading RUCSS.

Fixes #5401

## Type of change

Please delete options that are not relevant.

- [ ] Enhancement (non-breaking change which improves an existing functionality)

## Is the solution different from the one proposed during the grooming?

No.

## How Has This Been Tested?

- [ ] Automated Tests

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: inc/Engine/Optimization/RUCSS/Controller/UsedCSS.php [hunk 1]
```diff
@@ -238,6 +238,8 @@ public function treeshake( string $html ): string {
 				return $html;
 			}
 
+			do_action( 'rocket_preload_lock_url', $url );
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Controller/UsedCSS.php:219]: add docblock for action

File: inc/Engine/Optimization/RUCSS/Controller/UsedCSS.php [hunk 2]
```diff
@@ -611,7 +613,7 @@ public function check_job_status( int $id ) {
 			// Failure, check the retries number.
 			if ( $row_details->retries >= 3 ) {
 				Logger::debug( 'RUCSS: Job failed 3 times for url: ' . $row_details->url );
-
+				do_action( 'rocket_preload_unlock_url', $row_details->url );
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Controller/UsedCSS.php:637]: add docblock for action

File: inc/Engine/Preload/Subscriber.php [hunk 3]
```diff
@@ -109,7 +109,8 @@ public static function get_subscribed_events() {
 			'rocket_after_clean_terms'            => 'clean_urls',
 			'after_rocket_clean_domain'           => 'clean_full_cache',
 			'delete_post'                         => 'delete_post_preload_cache',
-			'pre_delete_term'                     => 'delete_term_preload_cache',
```
Comment (remyperona) [inline inc/Engine/Preload/Subscriber.php:115]: this callback should be preserved no?

**Review by remyperona — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@CrochetFeve0251 @Tabrisrp @piotrbak Please find the exploratory test note below:
With the following scenario, Clear and preload will change the status for locked URLs (the completed URLs while RUCSS still not generated) as the same as it was on the trunk
1- Preload / RUCSS are disabled and nothing in the database
2- Enable Preload and RUCSS
3- Once URL is completed as preload, it is locked and started to be added and  processed in the RUCSS
4- Clear and preload cache => all locked URLS in cache table with status completed and RUCSS not completed yet, Will change to pending, I think that locked URLs status should remain completed till RUCSS is done then it shall change to pending. What do you think? 
https://jmp.sh/Stw5NBv

**Comment by piotrbak** [comment]
@Mai-Saad I think that's okay.

**Comment by Mai-Saad** [comment]
@CrochetFeve0251 Can you please check the following scenario? (preload is done on locked URLs)
1- Enable preload and RUCSS
2- Completed cache URL is locked while being processed in RUCSS
3- Clear and preload + run `rocket_preload_process_pending` multiple times
4- The locked URL cache status will change to pending -> inprogress -> completed although it wasn't completed on RUCSS yet

**Comment by CrochetFeve0251** [comment]
> @CrochetFeve0251 Can you please check the following scenario? (preload is done on locked URLs) 1- Enable preload and RUCSS 2- Completed cache URL is locked while being processed in RUCSS 3- Clear and preload + run `rocket_preload_process_pending` multiple times 4- The locked URL cache status will change to pending -> inprogress -> completed although it wasn't completed on RUCSS yet

@Mai-Saad after testing I found that only the homepage have this problem. Do you have this only on that page?

**Comment by CrochetFeve0251** [comment]
@Mai-Saad  I think I got what you are saying: At the first loading of the cache the page is still going to complete in the preload.
This is due to the fact we need to generate the cache to launch RUCSS process and that trigger the complete status on the preload.

However if you run others preload request on the URL it won't  preload it. 

Should we prevent the first request to preload too @piotrbak ?

**Comment by piotrbak** [comment]
@CrochetFeve0251 The first request, when RUCSS is yet not triggered needs to happen. In this PR we cover the following scenario:
1. [domain.com/example](http://domain.com/example) is preloaded
2. It's added to the Used CSS table
3. It's locked until the Used CSS is there
4. We clear [domain.com/example](http://domain.com/example) cache
5. **It'll not be preloaded till Used CSS it there**
6. Used CSS appears in the database, the lock is set to 0
7. The URL will be preloaded

**Review by remyperona — APPROVED** [review]
There is a conflict to fix in the test class

**Review by Mai-Saad — APPROVED** [review]
Working as expected

**Pull request merged**